src/disposable/AsyncDisposable.ts

```diff
@@ -0,0 +1,42 @@
+import { IAsyncDisposable } from './types';
+
+export function createAsyncDisposable(func: () => void | Promise<void>): IAsyncDisposable {
+  return {
+    dispose: func,
+  };
+}
+
+export class AsyncDisposableCollection implements IAsyncDisposable {
+  protected readonly disposables: IAsyncDisposable[] = [];
+
+  get disposed(): boolean {
+    return this.disposables.length === 0;
+  }
+
+  public async dispose(): Promise<void> {
+    if (this.disposed) return;
+
+    while (!this.disposed) {
+      await this.disposables[this.disposables.length - 1].dispose();
+    }
+  }
+
+  public push(disposable: IAsyncDisposable): IAsyncDisposable {
```

Comment:
What about `pushAll`? I use it quite often.

Reply:
Feels like we could have a small check here to prevent the same instance from being added more than once.
I know `Disposable` does not have it, but we could really have it.

```
if (this.disposables.includes(disposable)) {
  return; // or perhaps throw?
}
```
At the very least, we should cover such a scenario with tests, as the way it's implemented in Disposable is different from what we have here (I mean the pop() call before disposing in Disposable)

Reply:
> What about `pushAll`? I use it quite often.

Ah. I forgot it, originally I was using `extends` so it just inherited the original `pushAll`. I'll add it back.

Reply:
> Feels like we could have a small check here to prevent the same instance from being added more than once.
> I know `Disposable` does not have it, but we could really have it.

🤔 We should probably add it to both, what do you think?

Reply:
> to prevent the same instance from being added more than once.

OR... should we use a Set? It actually makes the code a lot cleaner.